src/storage-client/src/storage_collections.rs

```diff
@@ -1126,6 +1161,80 @@ where
+        // we operate.
+        let collections_to_upgrade = {
+            let self_collections = self.collections.lock().expect("lock poisoned");
+
+            // Check if we're already switched over.
+            if !self.read_only.load(Ordering::SeqCst) {
+                // Already done!
+                return;
+            }
+
+            // Pull a copy of all collections whose leased handle we have to
+            // "upgrade" to a critical handle.
+            let mut collections_to_upgrade = Vec::new();
+
+            for (id, collection) in self_collections.iter() {
+                collections_to_upgrade.push((
+                    *id,
+                    collection.collection_metadata.data_shard.clone(),
+                    collection.read_capabilities.frontier().to_owned(),
+                ));
+            }
+
+            // Record that we're out of read-only mode.
+            self.read_only.store(false, Ordering::SeqCst);
+
+            // And only then allow adding collections again! This will make sure
+            // that any collections registered after this time will already have
+            // critical since handles.
+            drop(self_collections);
+
+            collections_to_upgrade
+        };
+
+        let persist_client = self
+            .persist
+            .open(self.persist_location.clone())
+            .await
+            .unwrap();
+
+        for (id, data_shard, since) in collections_to_upgrade {
+            let since_handle = self
+                .open_critical_handle(&id, data_shard, Some(&since), &persist_client)
+                .await;
+
+            // NOTE(aljoscha): This can happen when the "active" `environmentd`
+            // keeps driving forward the critical handle. If we wanted this to
+            // be guaranteed to work, we'd need a mechanism from persist that
+            // allows turning back a critical handle's since to a since that is
+            // held by a leased reader. In theory, this is possible, because the
+            // overall shard since is being held back by the leased handle that
+            // we hold here.
+            //
+            // We *shouldn't* invest into that for now, because the approach we
+            // use for zero-downtime upgrades does not require transitioning out
+            // of read-only mode within the process. We instead halt the new
+            // `environmentd` and let it come up fresh in read-write mode, at
+            // which time it will take over the critical since handles right at
+            // the beginning, as per the usual bootstrap process.
+            if PartialOrder::less_than(&since, since_handle.since()) {
+                panic!(
+                    "Tried upgrading leased handle of collection {id} to a critical handle, but the desired since ({:?}) is not beyond the critical since ({:?})",
+                    since, since_handle.since());
+            }
```

Comment:
I'm a bit worried that in a year our opinion about crash-and-burn being preferred will have changed and we try to write this code but forget to fix it. Could we ensure this doesn't happen? I think we could just log an error above when we enter it. Alternatively, do we need to include this code at all when we expect it to be dead code? Is it useful for testing?

Reply:
I found it useful for testing and it felt easier to write it now that I did the other read-only bits. But you're judgment might also be right here.

Re panic or logging: the panic at least makes it obvious that we're in a wrong state. If we just log and continue, we would continue in a potentially illegal state where we say that we can hold back the frontier using a critical handle to what the adapter needs from the controller, but we really can't. That seems more scary.

Reply:
If it's useful I'm fine with leaving it in.

> Re panic or logging: the panic at least makes it obvious that we're in a wrong state.

Sorry, I was not suggesting replacing the panic with a log! The panic is important for the reason you mentioned. But it's also not guaranteed to occur every time we enter this code, so my worry is that we might only trip it in production when it's too late. To avoid this we could have an error log ("you are running code that's not production-ready") further up at a place that _is_ guaranteed to be executed by the graceful switchover.

Reply:
ah yes! I'll add that

Reply:
I'm also worried about this code and if it stays I think it must be made to panic unless some test-only flag is set.